doc/testing.md

```diff
@@ -0,0 +1,158 @@
+# wasm is not happy with the instrumentation
+export SKIP_BUILD_WASM=true
+export BUILD_DUMMY_WASM_BINARY=true
+# the actully collected coverage data
+export LLVM_PROFILE_FILE="llvmcoveragedata-%p-%m.profraw"
+# build wasm without instrumentation
+export WASM_TARGET_DIRECTORY=/tmp/wasm
+cargo +nightly build
+# required rust flags
+export RUSTFLAGS="-Zinstrument-coverage"
+# assure target dir is clean
+cargo clean
+# run tests to get coverage data
+cargo +nightly test --all
+
+# create the *html* report out of all the test binaries
+# mostly useful for local inspection
+grcov . --binary-path ./target/debug -s . -t html --branch --ignore-not-existing -o ./coverage/
+miniserve -r ./coverage
+
+# create a *codecov* compatible report
+grcov . --binary-path ./target/debug/ -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov.info
+```
+
+The test coverage in `lcov` can the be published to <codecov.io>.
+
+```sh
+bash <(curl -s https://codecov.io/bash) -f lcov.info
+```
+
+or just printed as part of the PR using a gh action i.e. [jest-lcov-reporter](https://github.com/marketplace/actions/jest-lcov-reporter).
+
+For full examples on how to use [grcov /w polkadot specifics see the gh repo](https://github.com/mozilla/grcov#coverallscodecov-output).
+
+## Fuzzing
+
+Fuzzing is an approach to verify correctness against arbitrary or partially structured inputs.
+
+Currently implemented fuzzing targets:
+
+* `erasure-coding`
+* `bridges/storage-proof`
+
+The tooling of choice here is `honggfuzz-rs` as it allows _fastest_ coverage according to "some paper" which is a positive feature when run as part of PRs.
+
+Fuzzing is generally not applicable for data secured by cryptographic hashes or signatures. Either the input has to be specifically crafted, such that the discarded input
+percentage stays in an acceptable range.
+System level fuzzing is hence simply not feasible due to the amount of state that is required.
+
+Other candidates to implement fuzzing are:
+
+* `rpc`
+* ...
+
+## Writing small scope integration tests with preconfigured workers
+
+Requirements:
+
+* spawn nodes with preconfigured behaviours
+* allow multiple types of configuration to be specified
+* allow extensability via external crates
+* ...
+
+---
+
+
+## Implementation of different behaviour strain nodes.
+
+### Goals
+
+The main goals are is to allow creating a test node which
+exhibits a certain behaviour by utilizing a subset of _wrapped_ or _replaced_ subsystems easily.
+The runtime must not matter at all for these tests and should be simplistic.
+The execution must be fast, this mostly means to assure a close to zero network latency as
+well as shorting the block time and epoch times down to a few `100ms` and a few dozend blocks per epoch.
+
+### Approach
+
+`AllSubsystems` is an intermediate mocking type. As such it is a prime target for modification.
+`AllSubsystemsGen` is a proc-macro that should be extended as needed for per subsystem specific
```

Comment:
I think proc-macro would also be interesting, for the purpose of extracting the `Overseer` to be a more general utility, which can be used by other projects.

Reply:
Yeah, that'd be another option and certainly more useful. For the testing story, this won't make much of a difference impl wise.